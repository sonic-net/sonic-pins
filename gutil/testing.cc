// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include "gutil/testing.h"

#include "absl/strings/ascii.h"

namespace gutil {

// Implementation taken from protobuf descriptor library.
std::string SnakeCaseToCamelCase(absl::string_view input, bool lower_first) {
  bool capitalize_next = true;
  std::string result;
  result.reserve(input.size());

  for (char character : input) {
    if (character == '_') {
      capitalize_next = true;
    } else if (capitalize_next) {
      result.push_back(absl::ascii_toupper(character));
      capitalize_next = false;
    } else {
      result.push_back(character);
    }
  }

  // Lower-case the first letter.
  if (lower_first && !result.empty()) {
    result[0] = absl::ascii_tolower(result[0]);
  }

  return result;
}

}  // namespace gutil
