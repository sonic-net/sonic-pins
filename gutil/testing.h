// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef GUTIL_TESTING_H
#define GUTIL_TESTING_H

#include "absl/status/status.h"
#include "absl/strings/string_view.h"
#include "glog/logging.h"
#include "gutil/proto.h"
#include "gutil/status.h"

// Crash if `status` is not okay. Only use in tests.
#define CHECK_OK(val) CHECK_EQ(::absl::OkStatus(), (val))

namespace gutil {

// Parses a protobuf from a string, and crashes if parsing failed. Only use in
// tests.
template <typename T>
T ParseProtoOrDie(absl::string_view proto_string) {
  T message;
  CHECK_OK(ReadProtoFromString(proto_string, &message));
  return message;
}

// Parses a protobuf from a file, and crashes if parsing failed. Only use in
// tests.
template <typename T>
T ParseProtoFileOrDie(absl::string_view proto_file) {
  T message;
  CHECK_OK(ReadProtoFromFile(proto_file, &message));
  return message;
}

// Takes a snake_case string and returns a CamelCase string. If `lower_first` is
// set, the first character will be lowercase (if a letter) and otherwise it
// will be uppercase.
// Used to e.g. convert snake case strings to GTEST compatible test names.
std::string SnakeCaseToCamelCase(absl::string_view input,
                                 bool lower_first = false);

}  // namespace gutil

#endif  // GUTIL_TESTING_H
