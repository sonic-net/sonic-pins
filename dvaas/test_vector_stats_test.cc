// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dvaas/test_vector_stats.h"

#include <iostream>

#include "dvaas/test_vector.pb.h"
#include "gtest/gtest.h"
#include "gutil/testing.h"

namespace dvaas {
namespace {

PacketTestOutcomes GetPacketTestOutcomes() {
  return gutil::ParseProtoOrDie<PacketTestOutcomes>(R"pb(
    # Correct drop.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic drop.
          acceptable_outputs {
            packets: []
            packet_ins: []
          }
        }
        actual_output {
          packets: []
          packet_ins: []
        }
      }
      # Passed.
      test_result {}
    }

    # Incorrect forward.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic forward.
          acceptable_outputs {
            packets {}
            packet_ins: []
          }
        }
        actual_output {
          packets {}
          packet_ins: []
        }
      }
      # Failed.
      test_result {
        failure {
          # Failure is always deterministic.
          reproducibility_rate: 1.0
        }
      }
    }

    # Correct forward.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic forward.
          acceptable_outputs {
            packets {}
            packet_ins: []
          }
        }
        actual_output {
          packets {}
          packet_ins: []
        }
      }
      # Passed.
      test_result {}
    }

    # Correct multi forward.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic multi forward.
          acceptable_outputs {
            packets {}
            packets {}
            packets {}
            packet_ins: []
          }
        }
        actual_output {
          packets {}
          packets {}
          packets {}
          packet_ins: []
        }
      }
      # Passed.
      test_result {}
    }

    # Incorrect punt.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic trap.
          acceptable_outputs {
            packets: []
            packet_ins {}
          }
        }
        actual_output {
          packets: []
          packet_ins: {}
        }
      }
      # Failed.
      test_result {
        failure {
          # Failure is non-deterministic.
          reproducibility_rate: 0.0
        }
      }
    }

    # Incorrect punt with no reproducibility rate.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic trap.
          acceptable_outputs {
            packets: []
            packet_ins {}
          }
        }
        actual_output {
          packets: []
          packet_ins: {}
        }
      }
      # Failed.
      test_result { failure {} }
    }

    # Correct forward & copy.
    outcomes {
      test_run {
        test_vector {
          input {}
          # Deterministic forward & copy.
          acceptable_outputs {
            packets {}
            packet_ins {}
          }
        }
        actual_output {
          packets {}
          packet_ins: {}
        }
      }
      # Passed.
      test_result {
          # No reproducibility rate is given.
      }
    }
  )pb");
}

TEST(TestVectorStatsGoldenTest,
     ComputeTestVectorStatsAndExplainTestVectorStats) {
  PacketTestOutcomes outcomes = GetPacketTestOutcomes();
  TestVectorStats stats = ComputeTestVectorStats(outcomes);
  std::cout << ExplainTestVectorStats(stats);
}
}  // namespace
}  // namespace dvaas
