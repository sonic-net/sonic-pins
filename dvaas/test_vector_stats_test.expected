57.14% of 7 test vectors passed
100.00% of 7 test vectors produced the correct number and type of output packets
4 test vectors forwarded, producing 6 forwarded output packets
3 test vectors punted, producing 3 punted output packets
1 test vectors produced no output packets
None of 2 test vectors attempted had deterministically reproducible failures
57.14% of 7 test vectors passed
100.00% of 7 test vectors produced the correct number and type of output packets
4 test vectors forwarded, producing 6 forwarded output packets
3 test vectors punted, producing 3 punted output packets
1 test vectors produced no output packets
All of 2 test vectors attempted had deterministically reproducible failures
57.14% of 7 test vectors passed
100.00% of 7 test vectors produced the correct number and type of output packets
4 test vectors forwarded, producing 6 forwarded output packets
3 test vectors punted, producing 3 punted output packets
1 test vectors produced no output packets
